### ISSUE TITLE
Bug: Browsers don't respect setting preload to "none" in wavesurfer.load()

### ISSUE THREAD

**Issue by SeanBannister** [katspaugh/wavesurfer.js#973]
Demo at http://codepen.io/anon/pen/QdVxbe

Preload is set to none:
`wavesurfer.load('https://wavesurfer-js.org/example/split-channels/stereo.mp3', peaks, 'none');`
But if you check in "developer tools" the file is clearly loading on page load before the file is played. The DOM shows `preload="none"` so it is adding the attrbitue. 

I understand that `preload="none"` is only a suggestion to the browser but if I manually add the <audio> element to the page it respects it:
`<audio controls preload="none" src="https://wavesurfer-js.org/example/split-channels/stereo.mp3">`

Tested under Chrome 55.0.2883.87 and Firefox 51.0.1.

If you can give me an idea what might be causing this I'll try to fix it.

@X-Raym

**Comment by X-Raym** [comment]
@SeanBannister I use `preload="metadata"` and it seems to work as expected, so why not `preload="none"`... I don't have any idea.

**Comment by SeanBannister** [comment]
@X-Raym I just edited my example to use `preload="metadata"` and I'm definitely getting preloading, have a look: http://codepen.io/SeanBannister/pen/GrXBEp

**Comment by X-Raym** [comment]
@SeanBannister I did a some experimentation with my old codepens but I still don't have any hint about the situation. :(

**Comment by entonbiba** [comment]
@SeanBannister The problem is the `media.load();` that is being called in `_load: function (media, peaks) {` section.

With preload set to none it works if we check for 'none' in the _load function, but peaks will not load and audio file will not download which is the expected result.
![image](https://cloud.githubusercontent.com/assets/5193884/22764185/0ce1f77a-ee37-11e6-8da1-e2a3d2710dad.png)



You can make it work by loading an empty 0-1 sec audio file. After the user clicks the play button; it would re-load the .mp3 file with peaks again. It is still loading the audio file to render the peaks but it's 1 second vs the full audio file.

Preview:
http://codepen.io/entonbiba/pen/xgyGdr

![image](https://cloud.githubusercontent.com/assets/5193884/22764223/463ee780-ee37-11e6-980f-b1a35acf9ee5.png)

**Comment by mspae** [comment]
What's the status on this? @entonbiba @X-Raym @SeanBannister

**Comment by entonbiba** [comment]
@mspae the only way I see it working would be to somehow load a base64 0.01 second audio instead of the audio file itself when preload set to none. This would have to be included in the library though.

**Comment by X-Raym** [comment]
@mspae I get few reports from users who told me that full audio file is download, despite the fact peaks are pre-generated and that prelaod is set to metadata. :S

@entonbiba The audio files are loading because wavesurfer render HTML <audio> tag on the page. Maybe the trick would be to not render <audio> tag until a custom event happen (like clicking on play button or the waveform itself)... ? I think this could work, don't you ? (sure, it would need custom events and a bit of rewriting).

**Comment by mspae** [comment]
There must be another solution than rendering a faux media file. preload obviously does work without wavesurfer so there is something wrong with the implementation.

What about this:
- We log a warning when peaks aren't set and preload is set to none (telling the user that preload will be ignored in this case and that he must pass peaks) – also we document this by updating the jsdoc signature of the `ws.load` method.
- We check if preload is set to none before calling the `load()` function in `_load`
- The only thing remaining: how to actually the audio when we do need it.

**Comment by X-Raym** [comment]
@mspae just like you I think that a solution which doesnt involves dummy audio file will be better. Cause imagine you have 50 players on page... or if you have 1000 users per day. 0.01 secondes may not be heavy file size but it will be as many http requests. Not efficient.

@entonbiba do you think one of our propisition make sense here ?

**Comment by entonbiba** [comment]
@SeanBannister @X-Raym yea but will we be checking this only when using MediaElement backend or should we also implement the check for webaudio as well?

What about displaying an overlay with the text "Load Audio" which would work when clicked by the user only if preload option is set to none?

**Comment by X-Raym** [comment]
@entonbiba I think a custom event will be better to let the user manage the integration.
If we use precached peak, then no need to display anything :)

**Comment by dtbaker** [comment]
Anyone figure out how to disable pre-loading? 

I've got 30 mp3's on the page and don't want them to all load at once (pre-calculated waveform data is there)

**Comment by mspae** [comment]
@dtbaker  I'm currently putting together a PR: So far I've added informative warning messages if the preload attribute was set but will be ignored (e.g. `load` was called with a media element, not a string (as `url`), no peaks exist or the backend isn't `MediaElement`)

This doesn't  solve the actual loading issue (audio tracks appear to load even though the preload attribute was set) – I've been reading through the W3C specification: The relevant part is this: https://dev.w3.org/html5/spec-preview/media-elements.html#concept-media-load-resource

If I understand correctly the resource fetching algorithm, which is triggered by `media.load` loads the audio. The preload attribute is just a guideline but the exact behaviour (i.e. what to load) is up to the client to determine (depending on file size, network connection etc.)

>  The preload attribute is intended to provide a hint to the user agent about what the author thinks will lead to the best user experience. The attribute may be ignored altogether, for example based on explicit user preferences or based on the available connectivity.

[(source)](https://dev.w3.org/html5/spec-preview/media-elements.html#attr-media-preload)

This would mean we don't have absolute control over what actually happens and setting the correct preload attribute is the best we can do. I haven't been able to find a lot information about this issue. If anyone knows more, or is willing to investigate that would be awesome!

**Comment by mspae** [comment]
Also:

> HTMLMediaElement.load()
Resets the media element and restarts the media resource. Any pending events are discarded. How much media data is fetched is still affected by the preload attribute. This method can be useful for releasing resources after any src attribute and source element descendants have been removed. Otherwise, it is usually unnecessary to use this method, unless required to rescan source element children after dynamic changes.

[(source)](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement#Methods)

**Comment by X-Raym** [comment]
@mspae If preload=none is not respected by some browers, dont you think that a solution to prevent preload could be to not inject the audio tag at wavesurfer object init, but only after a custom event (like pressing on play button) ?
so preload: none would be a wavesurfer options attribute for its initiliaziation which would mean : dont inject audio html tag untill another event tell you so.
Just an idea.

**Comment by mspae** [comment]
@X-Raym I think it's a good idea. :)

I think for this feature we shouldn't use the `preload` parameter of the `load` function, but maybe a wavesurfer parameter called `lazyload` (default=`false`, works only with `MediaElement` backend)

Also we will need to explicitly pass the duration of the loaded audio file in order to calculate the correct `minPxPerSec` to render the wave. (Probably the media element equivalent of the "duration" part of https://github.com/katspaugh/wavesurfer.js/pull/1210 – maybe we can make a PR which adds a duration parameter for both backends for consistency's sake)

I propose you open a new issue for this feature in order to cleanly seperate this from the "preload" issue (which I would consider fixed when #1236 is merged)

**Comment by X-Raym** [comment]
@mspae Thanks for taking a look at this problem :)

Having exact duration passed is not a problem I think, for eg in case of pre-loaded peaks (the case where preload=none is useful), JSON peaks file can contains an entry for the peaks data, and one another the the audio duration.

But if length need to be exact, there is also this issue to look at: https://github.com/katspaugh/wavesurfer.js/issues/1223

I'll open a new issue 👍

**Comment by mspae** [comment]
Yes, I agree: In cases where this is useful (when you have peaks) you also know the length in advance and can provide wavesurfer with it.

I meant this commit as a reference how this could work (although this only works for the web audio backend currently) → https://github.com/katspaugh/wavesurfer.js/pull/1210/commits/1025d7263e32f087a3404012dbf0301e5f95b383

**Comment by X-Raym** [comment]
@mspae FR created : https://github.com/katspaugh/wavesurfer.js/issues/1237 !
